lib/semantics/assignment.cc

```diff
@@ -25,9 +26,218 @@
 #include "../parser/parse-tree.h"
 #include <optional>
 #include <set>
+#include <type_traits>
 
 using namespace Fortran::parser::literals;
 
+namespace Fortran::evaluate {
+
+template<typename A>
+void CheckPointerAssignment(parser::ContextualMessages &messages,
+    const IntrinsicProcTable &, const Symbol &symbol, const A &) {
+  // Default catch-all when RHS of pointer assignment isn't recognized
+  messages.Say("Pointer target assigned to '%s' must be a designator or "
+               "a call to a pointer-valued function"_err_en_US,
+      symbol.name().ToString().c_str());
+}
+
+void CheckPointerAssignment(parser::ContextualMessages &messages,
+    const IntrinsicProcTable &intrinsics, const Symbol &lhs,
+    const NullPointer &f) {
+  // LHS = NULL() without MOLD=; this is always fine
+}
+
+template<typename T>
+void CheckPointerAssignment(parser::ContextualMessages &messages,
+    const IntrinsicProcTable &intrinsics, const Symbol &lhs,
+    const FunctionRef<T> &f) {
+  const Symbol *ultimate{nullptr};
+  std::string funcName;
+  if (const auto *symbol{f.proc().GetSymbol()}) {
+    funcName = symbol->name().ToString();
+    ultimate = &symbol->GetUltimate();
+  } else if (const auto *intrinsic{f.proc().GetSpecificIntrinsic()}) {
+    funcName = intrinsic->name;
+  }
+  if (auto proc{Characterize(f.proc(), intrinsics)}) {
+    std::optional<parser::MessageFixedText> error;
+    if (const auto &funcResult{proc->functionResult}) {
+      const auto *frProc{funcResult->IsProcedurePointer()};
+      if (IsProcedurePointer(lhs)) {
+        // Shouldn't be here in this function unless lhs
+        // is an object pointer.
+        error = "Procedure pointer '%s' assigned with result of "
+                "reference to function '%s' that does not return a "
+                "procedure pointer"_err_en_US;
```

Comment:
"Assigned with" sounds odd to me. How about "associated with"?

Reply:
"was assigned the result" may be even better